fda-model/stageDamage/ImpactAreaStageDamage.cs

```diff
@@ -188,79 +190,221 @@ private PairedData CreateStageFrequency()
+                    List<double> allStagesAtIndexLocation = new List<double>();
+
+                    //There will be one ConsequenceDistributionResults object for each stage in the stage-damage function
+                    //Each ConsequenceDistributionResults object holds a ConsequenceDistributionResult for each asset cat
+                    List<ConsequenceDistributionResults> consequenceDistributionResults = new List<ConsequenceDistributionResults>();
+
+                    (Inventory, List<float[]>) inventoryAndWaterTupled = _inventory.GetInventoryAndWaterTrimmedToDamageCategory(damageCategory, wsesAtEachStructureByProfile.Item2);
+
+                    //Run a first pass to generate sufficient sample size to generate good histograms
+                    bool isFirstPass = true;
+                    ComputeDamageWithUncertaintyAllCoordinates(ref consequenceDistributionResults, ref allStagesAtIndexLocation, damageCategory, randomProvider, inventoryAndWaterTupled, wsesAtEachStructureByProfile.Item1, isFirstPass);
+                    bool functionIsNotConverged = IsTheFunctionNotConverged(consequenceDistributionResults);
+
+                    isFirstPass = false;
+
+                    while (functionIsNotConverged)
+                    {
+                        ComputeDamageWithUncertaintyAllCoordinates(ref consequenceDistributionResults, ref allStagesAtIndexLocation, damageCategory, randomProvider, inventoryAndWaterTupled, wsesAtEachStructureByProfile.Item1, isFirstPass);
+                        functionIsNotConverged = IsTheFunctionNotConverged(consequenceDistributionResults);
+                    }
+                    //there should be four UncertainPairedData objects - one for each asset cat of the given dam cat level compute 
+                    List<UncertainPairedData> tempResultsList = ConsequenceDistributionResults.ToUncertainPairedData(allStagesAtIndexLocation, consequenceDistributionResults, _ImpactAreaID);
+                    results.AddRange(tempResultsList);
+                }
+                return results;
+            }
+        }
+
+        private void ComputeDamageWithUncertaintyAllCoordinates(ref List<ConsequenceDistributionResults> consequenceDistributionResults, ref List<double> allStagesAtIndexLocation, string damageCategory, IProvideRandomNumbers randomProvider, (Inventory, List<float[]>) inventoryAndWaterTupled, List<double> profileProbabilities, bool isFirstPass)
+        {
+            List<Dictionary<string, List<double>>> assetCatDamagesAllCoordinates = new List<Dictionary<string, List<double>>>();
+            //we need to instantiate all stages at index location now 
+            //maybe we dont -- maybe we just use the stage frequency relationship like 
+            int iterations = convergenceCriteria.MinIterations;
+            //TODO: we need to track if is very first pass 
+            //"are dictionaries constructed?" if i = 0 then no 
+            bool dictionariesAreNotConstructed = true;
```

Comment:
Chatted about this in the office. It was a choice to do it this way, rather than have a more typical name with a negation (!). Not my preference, but I can respect the choice is inconsequential.